### PR TITLE
An overhaul of Nim's CI

### LINKED ISSUE #17325: CI fails on azure 32 bit linux Linux_i386: sudo apt-fast install fails
I'm seeing:
`nim-lang.Nim (packages Linux_i386) Failing after 58s — packages Linux_i386 failed`
in several unrelated PRs (eg https://github.com/nim-lang/Nim/pull/17257#issuecomment-796247652)

eg https://dev.azure.com/nim-lang/255dfe86-e590-40bb-a8a2-3c0295ebdeb1/_apis/build/builds/14075/logs/38

```
2021-03-10T21:10:24.0269335Z CC: stdlib_assertions.nim
2021-03-10T21:10:24.0288424Z CC: stdlib_dollars.nim
2021-03-10T21:10:24.1410956Z Error: execution of an external compiler program 'gcc -c  -w -fmax-errors=3  -I/home/vsts/work/1/s/lib -I/home/vsts/work/1/s -o /home/vsts/.cache/nim/koch_d/stdlib_assertions.nim.c.o /home/vsts/.cache/nim/koch_d/stdlib_assertions.nim.c' failed with exit code: 1
2021-03-10T21:10:24.1412125Z 
2021-03-10T21:10:24.1412666Z In file included from /usr/include/limits.h:25:0,
2021-03-10T21:10:24.1413573Z                  from /usr/lib/gcc/x86_64-linux-gnu/5/include-fixed/limits.h:168,
2021-03-10T21:10:24.1414497Z                  from /usr/lib/gcc/x86_64-linux-gnu/5/include-fixed/syslimits.h:7,
2021-03-10T21:10:24.1415407Z                  from /usr/lib/gcc/x86_64-linux-gnu/5/include-fixed/limits.h:34,
2021-03-10T21:10:24.1416047Z                  from /home/vsts/work/1/s/lib/nimbase.h:268,
2021-03-10T21:10:24.1416680Z                  from /home/vsts/.cache/nim/koch_d/stdlib_assertions.nim.c:9:
2021-03-10T21:10:24.1417570Z /usr/include/features.h:367:25: fatal error: sys/cdefs.h: No such file or directory
2021-03-10T21:10:24.1418154Z compilation terminated.
2021-03-10T21:10:24.1472789Z ##[error]Bash exited with code '1'.
2021-03-10T21:10:24.1576588Z ##[section]Finishing: Build koch
```

(refs https://github.com/nim-lang/Nim/issues/3123#issuecomment-122125096)
refs:
* sudo apt-get install g++-multilib https://stackoverflow.com/questions/51765069/fatal-error-sys-cdefs-h-no-such-file-or-directory-while-compiling-openssl-1-1
* libc6-dev-i386 https://itectec.com/ubuntu/ubuntu-fatal-error-sys-cdefs-h-no-such-file-or-directory/

### PR DESCRIPTION
**Pull request by alaviss**
Todo:
- [ ] Consistent environment
  - [x] Linux
    - [x] Node.js v12 (except i386, Node.js no longer has official support for this platform since v10)
    - [x] GCC 8.4.0 (as requested by @Araq)
  - [ ] Windows
    - [x] Node.js v12
    - [ ] GCC 8.4.0
- [ ] Split components into shell scripts, the `ci/` folder can be re-used for this. Also consider copying some code from nightlies.

Fixes #17325

### CODE REVIEW

**Comment by Varriount** [comment]
Depending on the complexity of the various tasks, you might want to consider using something like Python, Ruby, or Perl over Bash.

**Comment by timotheecour** [comment]
in https://github.com/nim-lang/Nim/pull/17359#issuecomment-797880252 I was able to make i386 work again using a much smaller patch (i could turn it into a cleaned up PR if needed) but 3 tests failed:

* tests/async/tasyncssl.nim
* tests/stdlib/tssl.nim
* tests/niminaction/Chapter8/sfml/sfml_test.nim

still though, isn't disabling 3 tests better than disabling all `nim js` tests on 32bit linux?

that said, a docker based approach might be more flexible than the way we've been doing via cross compilation (although it does serve as an example of cross compilation tested in CI, which is useful)

can't we have a `Dockerfile` in the sources instead of relying on an external docker image (which is hard/impossible to change in PR's, say when a new dependency is needed). At least github actions supports using Dockerfile.

## links
* https://github.com/nodejs/build/issues/885

**Comment by dom96** [comment]
> Depending on the complexity of the various tasks, you might want to consider using something like Python, Ruby, or Perl over Bash.

Better yet, consider using Nim. You can easily grab a Nim release that is known to work for the CI code.

File: azure-pipelines.yml [hunk 1]
```diff
@@ -50,145 +42,29 @@ jobs:
...
-        else
-          echo_run make -C csources -j $ncpu CC=gcc ucpu=$(CPU) koch=no
-        fi
+  strategy:
+    matrix:
+      amd64:
+        containerImage: ghcr.io/alaviss/nim-ci:sha-aa56233@sha256:43dbe085d041c95b5d39bcbc5e0c4e05d922d407ac8a8938d33064402a5a5f22
+        CPU: amd64
+      i386:
+        containerImage: ghcr.io/alaviss/nim-ci:sha-aa56233@sha256:7f6d204d43e213ff8621e9f0c9f2b788858b9f3788fed95ac62e1c821baa4476
```
Comment (juancarlospaco) [inline azure-pipelines.yml:57]: Everything to test the repo must live in the Nim organization IMHO.

**Comment by alaviss** [comment]
@timotheecour 

> still though, isn't disabling 3 tests better than disabling all `nim js` tests on 32bit linux?

It's mostly out of practical concerns. Node.js no longer provides official builds on 32-bit x86 Linux, and support for it has been moved to "experimental". I can add a recent Node.js to the image by building it from source, but I don't consider that worth the hassle, considering that no one in their right mind would run Node.js on 32-bit x86 Linux.

If we want 32-bit Node testing, there is Windows and my ARM CI in #16396.

> that said, a docker based approach might be more flexible than the way we've been doing via cross compilation (although it does serve as an example of cross compilation tested in CI, which is useful)

The way we "test" cross compilation is mostly a hack. `./koch boot` doesn't understand cross compilers at all, and we don't test the usage of `<arch>.<os>.<cfg>` configuration variables. For now I'd say that nightlies fulfills this role, given how it cross compile to x86_32, armv7 and arm64 (though we still don't "cross" compile but we use "native" cross compilers).

> can't we have a `Dockerfile` in the sources instead of relying on an external docker image (which is hard/impossible to change in PR's, say when a new dependency is needed). At least github actions supports using Dockerfile.

I can move the [alaviss/nim-ci](https://github.com/alaviss/nim-ci) repo to nim-lang/nim-ci.

The main reason for keeping it in a separated repo is so that we can develop further images there, not just for CI. Also, if you look at the GitHub Actions used to build my images, it takes 7 minutes to complete, and I don't want to shoulder that cost on the main CI.

@dom96 

> Better yet, consider using Nim. You can easily grab a Nim release that is known to work for the CI code.

That would only work if you're testing Windows, Linux or macOS. We test the BSDs too, and there are no prebuilts for them.

And most of the code we're dealing with is just environment setup. 80% of CI work is done in `koch` already.

File: azure-pipelines.yml (hunk 1 above)
Comment (ghost) [reply azure-pipelines.yml:57]: @juancarlospaco then maybe just move nim-ci under @nim-lang organization, no need to have it in the same repo

Comment (juancarlospaco) [reply azure-pipelines.yml:57]: True

**Pull request closed without merging**